code/modules/flufftext/Hallucination.dm

```diff
@@ -815,29 +935,53 @@ GLOBAL_LIST_INIT(hallucination_list, list(
 		if("beepsky")
 			target.playsound_local(source, 'sound/voice/beepsky/freeze.ogg', 35, 0)
 		if("mech")
-			var/mech_dir = pick(GLOB.cardinals)
-			for(var/i in 1 to rand(4,9))
-				if(prob(75))
-					target.playsound_local(source, 'sound/mecha/mechstep.ogg', 40, 1)
-					source = get_step(source, mech_dir)
-				else
-					target.playsound_local(source, 'sound/mecha/mechturn.ogg', 40, 1)
-					mech_dir = pick(GLOB.cardinals)
-				sleep(10)
+			new /datum/hallucination/mech_sounds(C, forced, sound_type)
 		//Deconstructing a wall
 		if("wall decon")
 			target.playsound_local(source, 'sound/items/welder.ogg', 50, 1)
-			sleep(105)
-			target.playsound_local(source, 'sound/items/welder2.ogg', 50, 1)
-			sleep(15)
-			target.playsound_local(source, 'sound/items/ratchet.ogg', 50, 1)
+			addtimer(CALLBACK(target, /mob/.proc/playsound_local, source, 'sound/items/welder2.ogg', 50, 1), 105)
+			addtimer(CALLBACK(target, /mob/.proc/playsound_local, source, 'sound/items/ratchet.ogg', 50, 1), 120)
 		//Hacking a door
 		if("door hack")
 			target.playsound_local(source, 'sound/items/screwdriver.ogg', 50, 1)
-			sleep(rand(40,80))
-			target.playsound_local(source, 'sound/machines/airlockforced.ogg', 30, 1)
+			addtimer(CALLBACK(target, /mob/.proc/playsound_local, source, 'sound/machines/airlockforced.ogg', 30, 1), rand(40, 80))
 	qdel(src)
 
+/datum/hallucination/mech_sounds
+	var/mech_dir
+	var/steps_left
+	var/next_action = 0
+
+/datum/hallucination/mech_sounds/New()
+	. = ..()
+	mech_dir = pick(GLOB.cardinals)
+	steps_left = rand(4, 9)
+	START_PROCESSING(SSfastprocess, src)
+
+/datum/hallucination/mech_sounds/process(delta_time)
+	var/turf/source = random_far_turf()
+
+	next_action -= delta_time
+	if (next_action > 0)
+		return
+
+	if(prob(75))
+		target.playsound_local(source, 'sound/mecha/mechstep.ogg', 40, 1)
+		source = get_step(source, mech_dir)
```

Comment:
you're not actually moving the source of the sound here, you'll need to set it on init instead of each process